### PR TITLE
[HACK] Turn Slack's invalid 332s into 331s instead

### PR DESCRIPTION
**Pull request by knewter**
This addresses #51.  I'm sure there should be asomething nicer where I'm
handling this - do you have any thoughts @bitwalker?

### CODE REVIEW

File: lib/exirc/utils.ex [hunk 1]
```diff
@@ -78,7 +78,17 @@ defmodule ExIrc.Utils do
...
       |> Enum.map(&trim_crlf/1)
       |> Enum.map(&List.to_string/1)
-    %{msg | args: args}
+
+    # Handle Slack's inappropriate RPL_TOPIC when they should send RPL_NOTOPIC:
+    # https://github.com/bitwalker/exirc/issues/51
+    case {msg, args} do
+      { %IrcMessage{cmd: "332"}, [_] } ->
+        # We shouldn't end up in the situation where we have a "332" but only one arg.  Turn this into a 331.
+        [channel, nick] = hd(msg.args)
```
Comment (bitwalker) [inline lib/exirc/utils.ex]: This can be moved into the pattern for the case clause, but I can take care of that after merging if time is tight.

**Comment by bitwalker** [comment]
This is indeed a dirty hack haha, but I'm not sure what the better way of handling it is, since we're deliberately working around a violation of the spec. That said, if I knew the error exirc produced when it exploded when you first encountered this, I might be able to come up with something that feels less dirty.

If it doesn't look like there are any good options, I'll just merge this as-is and refactor it later if the opportunity arises.

**Comment by knewter** [comment]
It was a NoClause error in the case statement here:
https://github.com/bitwalker/exirc/blob/master/lib/exirc/client.ex#L575

**Comment by bitwalker** [comment]
Hmm, what about adding a third clause there which only matches a single arg, and defaults `topic` to `"No topic is set"`? That case clause already handles one non-RFC compliant mutation (although in that case it's at least a reasonable mutation). 

I'm not sure what's better, to transform a non-conforming command to a conforming one, or to just handle the non-conformance better in the handler for that command. Both solutions seem reasonable, thoughts?

**Comment by knewter** [comment]
My thoughts were that there was no code to handle the 331 yet and
consequently this had less impact on end users - they haven't gotten a no
topic message before, maybe (???) it would have some weird frustration for
them.  Either seems fine to me.

**Comment by bitwalker** [comment]
Yeah I need to add a handler for the no topic message. I'll look at this closer today and apply a fix one way or the other.

**Comment by bitwalker** [comment]
@knewter I manually merged your changes (and made you the author of the commits) with some small tweaks. I think I found a typo bug in the pre-existing code in the utils module though, and while the tests are passing, there is the potential I missed something about why I wrote it that way to begin with. In any case, if you could test out master in your app and let me know if things look good, I'll cut a release if things look stable on your end.

**Comment by knewter** [comment]
@bitwalker it seems just fine now!  There's still a problem with slack's IRC bridge that only gets triggered at some particular number of users...really hope I get a good backtrace for that in the future.  I can't reproduce it now but the night you first cut this I could reliably.  It's related to them sending the list of users in multiple messages due to the message length requirements, but then the last list is empty which leads to an invalid message on their part again.  Hard to replicate from here atm sadly :-\

Anyway, so far everything's great and have no more crashes tonight with the irc gateway :)

**Comment by bitwalker** [comment]
Interesting, I'll play around and see if I can reproduce that scenario and find a way to handle it better, if you see it again and capture a trace that will be very useful too :)

**Pull request closed without merging**